src/Avalonia.Base/Layout/LayoutManager.cs

```diff
@@ -352,7 +352,7 @@ private bool RaiseEffectiveViewportChanged()
                         if (viewport != l.Viewport)
                         {
                             l.Listener.EffectiveViewportChanged(new EffectiveViewportChangedEventArgs(viewport));
-                            _effectiveViewportChangedListeners[i] = new EffectiveViewportChangedListener(l.Listener, viewport);
+                            l.Viewport = viewport;
```

Comment:
I did not try to salvage the "readonly" class attribute. This looked faster and safer.